aspects/schema.go

```diff
@@ -0,0 +1,328 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+type parser interface {
+	Schema
+
+	// parseConstraints parses constraints for a type defined as a JSON object.
+	// Shouldn't be used with non-object/map type definitions.
+	parseConstraints(json.RawMessage) error
```

Comment:
maybe I'm missing something but it seems to me that this could have the signature:

parseConstraints(schemaDef map[string]*json.RawMessage)

and avoid the double unmarshalling?